oneflow/api/cpp/framework/graph.cpp

```diff
@@ -0,0 +1,256 @@
+#include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
+#include "oneflow/api/python/job_build/job_build_and_infer.h"
+#include "oneflow/core/common/data_type.pb.h"
+#include "oneflow/core/common/global.h"
+#include "oneflow/core/common/hash_container.h"
+#include "oneflow/core/common/just.h"
+#include "oneflow/core/common/shape.h"
+#include "oneflow/core/common/symbol.h"
+#include "oneflow/core/eager/eager_oneflow.h"
+#include "oneflow/core/framework/device.h"
+#include "oneflow/core/framework/dtype.h"
+#include "oneflow/core/framework/multi_client_session_context.h"
+#include "oneflow/core/framework/nn_graph.h"
+#include "oneflow/core/framework/scope_util.h"
+#include "oneflow/core/framework/tensor.h"
+#include "oneflow/core/framework/tensor_tuple.h"
+#include "oneflow/core/functional/functional_api.yaml.h"
+#include "oneflow/core/graph/op_graph.h"
+#include "oneflow/core/job/job.pb.h"
+#include "oneflow/core/job/job_build_and_infer_ctx.h"
+#include "oneflow/core/job/job_build_and_infer_ctx_mgr.h"
+#include "oneflow/core/job/job_conf.cfg.h"
+#include "oneflow/core/job/job_conf.pb.h"
+#include "oneflow/core/job/job_ir.h"
+#include "oneflow/core/job/job_set.pb.h"
+#include "oneflow/core/job/lazy_mode.h"
+#include "oneflow/core/job/parallel_desc.h"
+#include "oneflow/core/job/scope.h"
+#include "oneflow/core/job/session.h"
+#include "oneflow/core/operator/interface_blob_conf.pb.h"
+#include "oneflow/core/operator/op_conf.pb.h"
+
+namespace oneflow_api {
+
+namespace of = oneflow;
+
+namespace {
+
+class CompileScope {
+ public:
+  CompileScope(const of::JobConfigProto& job_config, const std::shared_ptr<of::NNGraph>& graph_) {
+    std::shared_ptr<of::Scope> scope = of::MakeInitialScope(job_config).GetOrThrow();
+    of::ThreadLocalScopeStackPush(scope).GetOrThrow();
+    of::JobBuildAndInferCtx_Open(job_config.job_name()).GetOrThrow();
+    of::cfg::JobConfigProto job_config_cfg(job_config);
+    of::CurJobBuildAndInferCtx_SetJobConf(job_config_cfg).GetOrThrow();
+  }
+
+  ~CompileScope() {
+    of::JobBuildAndInferCtx_Close().GetOrThrow();
+    of::ThreadLocalScopeStackPop().GetOrThrow();
+  }
+
+ private:
+  of::LazyMode::Guard lazy_mode_enabled_guard{true};
+};
+
+std::shared_ptr<of::one::TensorTuple> ConvertToTensorTuple(
+    const std::vector<std::shared_ptr<of::one::Tensor>>& tensors) {
+  auto tensor_tuple = std::make_shared<of::one::TensorTuple>();
+  for (const auto& tensor : tensors) { tensor_tuple->emplace_back(tensor); }
+  return tensor_tuple;
+}
+
+}  // namespace
+
+Graph::Graph(const std::string& model_path, const Device& device) : device_(device) {
+  // TODO(zzk0): model_path is a directory, need to concatenate filename
+  // we need a mlir model name.
+  // of::LoadJobFromIR(&job_, model_path).GetOrThrow();
+
+  std::ifstream input("/home/zhouzekai/oneflow/oneflow/ir/test/saved_model/job.pb");
+  std::stringstream buffer;
+  buffer << input.rdbuf();
+  job_.ParseFromString(buffer.str());
+  input.close();
```

Comment:
```suggestion
  {
    std::ifstream input("/home/zhouzekai/oneflow/oneflow/ir/test/saved_model/job.pb");
    CHECK(job_.ParseFromIstream(&input));
  }
```
可以直接用 ParseFromIstream，并且要检查返回值；`input` 用完就失效了，所以限制它的生命周期